### PR TITLE
Adds option for Slider to snap to discrete values

### PR DESCRIPTION
**Pull request by jcalbert**
## PR Summary
Addresses #9129, allows the slider UI widget to snap to multiples of optional kwarg `valstep`.
 

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -271,9 +271,9 @@ class Slider(AxesWidget):
 
     Call :meth:`on_changed` to connect to the slider event
     """
-    def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt='%1.2f',
-                 closedmin=True, closedmax=True, slidermin=None,
-                 slidermax=None, dragging=True, **kwargs):
+    def __init__(self, ax, label, valmin, valmax, valstep=None, valinit=0.5,
+                 valfmt='%1.2f', closedmin=True, closedmax=True,
+                 slidermin=None, slidermax=None, dragging=True, **kwargs):
```
Comment (WeatherGod) [inline lib/matplotlib/widgets.py]: Don't change the order of the arguments in an existing function, even if they are keyword arguments.

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -368,6 +372,8 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt='%1.2f',
 
     def _value_in_bounds(self, val):
         """ Makes sure self.val is with given bounds."""
+        if self.valstep:
+            val = round(val/self.valstep)*self.valstep
```
Comment (WeatherGod) [inline lib/matplotlib/widgets.py]: should use `np.round()` to get consistent behavior in python 2.7 and py3k.

**Comment by tacaswell** [comment]
I am 50/50 on rushing this for 2.1.

File: lib/matplotlib/widgets.py (hunk 1 above)
Comment (jcalbert) [reply lib/matplotlib/widgets.py]: What's the proper place to put them, before kwargs?

Comment (WeatherGod) [reply lib/matplotlib/widgets.py]: yes

**Comment by jcalbert** [comment]
Before setting it in stone, a few questions about expected behavior:
1)  Would it be preferable for the little blue bar to track the cursor while dragging, and then snap back to an allowed value on button_release?
2)  For `valmin=.2, valmax=4.5, valstep=1.0` should the points go `[0.2, 1.2, 2.2, 3.2, 4.2, 4.5]` or `[0.2, 1.0, 2.0, 3.0, 4.0, 4.5]` ?

**Comment by WeatherGod** [comment]
Personally, I think it should track the cursor. It provides visual feedback to the user that the slider is active. However, is it possible to implement without it reporting non-valid numbers?

For the second point, that's a toughie. I think I like the first option better.

**Comment by jcalbert** [comment]
@WeatherGod I tried a version with the tracking cursor, but I don't think it 'feels' right.  Even with some tick marks as a guide, it feels like the UI is just being laggy.  You can try it [here](https://github.com/jcalbert/matplotlib/tree/smooth_discrete_slider).

File: doc/users/next_whats_new/2017-08-31_discrete-sliders.rst [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Slider UI widget can snap to discrete values
+--------------------------
```
Comment (dstansby) [inline doc/users/next_whats_new/2017-08-31_discrete-sliders.rst]: These dashes should be exactly the same length as the title

**Comment by jcalbert** [comment]
^BTW, I noticed that the slider widget doesn't implement blitting like most of the other widgets do.  Is that something worth looking into?

**Comment by WeatherGod** [comment]
Blitting makes sense for things like drawing a rubber band, but I doubt
it'll really be useful for sliders. If people find that there is room for
performance improvements, then possibly. For now, I'd leave that off.

**Pull request merged**